### PR TITLE
chore(docs): add example yamls and docs for cstor upgrades/migration

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

This PR adds sample upgrade job yamls for cstor CSPC and Volume upgrades. This PR also adds documentation for migration of cstor pools and volumes.

### CODE REVIEW

**Comment by kmova** [comment]
@shubham14bajpai - If there are no further updates required, please remove the hold-merge label.

File: docs/migration.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+ - The cstor-operator should be installed with version `1.12.0` or above. 
+    You can install the correct version of cstor-operator from [charts](https://github.com/openebs/charts/tree/gh-pages). Get the cstor-operator yaml within the correct versioned folder and install.
+ - The SPC pool version that needs to migrated should be `1.12.0` or above. 
+    You can use the following command to verify the version of SPC. The output should be the current version of SPC. 
+    ```sh
+    kubectl get spc <spc-name> -o jsonpath="{.versionDetails.status.current}"
+    ```
+    For example if spc-name is `sparse-claim` then the output will be:
+    ```sh
+    $ kubspc sparse-claim -o jsonpath="{.versionDetails.status.current}"
```
Comment (nsathyaseelan) [inline docs/migration.md:45]: @shubham14bajpai  change the command `kubectl get spc`

File: docs/migration.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        image: openebs/migrate-amd64:1.12.0
+      restartPolicy: Never
+---
+```
+
+You can get the above yaml from [here](../examples/migrate/spc-migration.yaml).
+
+The status of the job can be verified by looking at the logs of the job pod. To get the job pod use the command:
+```sh
+$ k get pods -l job-name=migrate-spc-sparse-claim
```
Comment (nsathyaseelan) [inline docs/migration.md:105]: `kubectl`

File: docs/migration.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+```
+
+## CStor External Provisioned volumes to CStor CSI volumes
+
+### Prerequisites
+
+Before migrating the volumes make sure the following prerequisites are taken care of:
+
+ - The first two prerequisites for [pool](#spc-pools-to-cspc-pools) are required for volumes as well.
+ - The csi-operator should be installed with version 1.12.0 or above. You can install the correct version of csi-operator from charts. Get the csi-operator yaml within the correct versioned folder and install. The version should be same as the cstor-operator installed.
```
Comment (nsathyaseelan) [inline docs/migration.md:130]: can we give the link for csi-operator here?

File: docs/migration.md [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+I0520 10:00:53.459935       1 pool.go:102] Creating equivalent cspc for spc sparse-claim
+I0520 10:00:59.221476       1 pool.go:367] Updating bdc bdc-17771e05-1c55-43ef-abb3-39dcc36472d4 with cspc ownerRef.
+I0520 10:00:59.233294       1 pool.go:367] Updating bdc bdc-d9e6b0ff-c951-4081-903e-56bb71900adf with cspc ownerRef.
+I0520 10:00:59.510423       1 pool.go:221] Migrating csp sparse-claim-dzr2 to cspi sparse-claim-mbo2
+I0520 10:00:59.510448       1 pool.go:282] Scaling down deployemnt sparse-claim-mbo2
+I0520 10:01:25.674512       1 pool.go:395] Updating cvr pvc-9cf5a405-12c0-4522-b031-7816425f443f-sparse-claim-mbo2 with cspi sparse-claim-dzr2 info.
+I0520 10:01:25.798889       1 pool.go:80] Successfully migrated spc sparse-claim to cspc
+```
+
+## CStor External Provisioned volumes to CStor CSI volumes
```
Comment (nsathyaseelan) [inline docs/migration.md:123]: cStor or CStor in docs we are following cStor

File: docs/migration.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+I0713 12:51:50.309624       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: 
+I0713 12:52:00.319407       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: 
+I0713 12:53:11.175510       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: Init
+I0713 12:53:21.177993       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: Init
+I0713 12:53:31.242181       1 volume.go:956] Patching the target svc with cvc owner ref
+I0713 12:53:31.336819       1 volume.go:1029] Cleaning up old volume resources
+I0713 12:53:31.714056       1 cstor_volume.go:80] Successfully migrated volume pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d, scale up the application to verify the migration
+```
+**Note:** If target affinity was set old volume the target pod will go into `pending` state after the migration is completed. Once the application is scaled up the target pod should automatically reschedule to the same node as application.
```
Comment (nsathyaseelan) [inline docs/migration.md:211]: If target affinity was set to old volume

File: docs/upgrade.md [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+```
+
+## cStor CSI volumes
+
+These instructions will guide you through the process of upgrading cStor CSI volumes from `1.11.0` or later to a newer release up to `1.12.0`.
+
+### Prerequisites
+
+Before upgrading the volumes make sure the following prerequisites are taken care of:
+
```
Comment (nsathyaseelan) [inline docs/upgrade.md:138]: @shubham14bajpai scaledown of application is recommended or not?

Comment (shubham14bajpai) [reply docs/upgrade.md:138]: Added a note

File: docs/migration.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+I0713 12:51:50.309624       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: 
+I0713 12:52:00.319407       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: 
+I0713 12:53:11.175510       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: Init
+I0713 12:53:21.177993       1 volume.go:951] Waiting for cv pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d to come to Healthy state, got: Init
+I0713 12:53:31.242181       1 volume.go:956] Patching the target svc with cvc owner ref
+I0713 12:53:31.336819       1 volume.go:1029] Cleaning up old volume resources
+I0713 12:53:31.714056       1 cstor_volume.go:80] Successfully migrated volume pvc-7ac10812-cc83-4fc5-a2e0-7d24f785e93d, scale up the application to verify the migration
+```
+**Note:** If target affinity was set to the old volume, the target pod will go into `pending` state after the migration is completed. Once the application is scaled up the target pod should automatically reschedule to the same node as application.
```
Comment (mittachaitu) [inline docs/migration.md:213]: We need to mention this [point](https://github.com/openebs/upgrade/pull/15#pullrequestreview-446437868) also.

File: docs/upgrade.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
+# Upgrade OpenEBS
+
+This document describes the steps for upgrading the following OpenEBS reources:
+
+- [CSPC pools](#cspc-pools)
+- [cStor CSI volumes](#cstor-csi-volumes)
+
+## CSPC pools
+
+These instructions will guide you through the process of upgrading cStor CSPC pools from `1.11.0` or later to a newer release up to `1.12.0`.
```
Comment (mittachaitu) [inline docs/upgrade.md:10]: Q: Is there any specific reason for not supporting CSPC pool from `1.10.0` to the new Version?

File: examples/migrate/cstor-volume-migration.yaml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        env:
+        - name: OPENEBS_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        tty: true
+        # the version of the image should be same the 
+        # version of cstor-operator installed.
+        image: openebs/migrate-amd64:1.12.0
+      restartPolicy: Never
```
Comment (mittachaitu) [inline examples/migrate/cstor-volume-migration.yaml:36]: Q: Is there any specif reason for not having `OnFailure`?

File: examples/migrate/spc-migration.yaml [hunk 10]
```diff
@@ -0,0 +1,40 @@
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  # the name can be of the form migrate-<spc-name>
+  name: migrate-spc-sparse-claim
+  namespace: openebs
+spec:
+  backoffLimit: 0
```
Comment (mittachaitu) [inline examples/migrate/spc-migration.yaml:9]: Q: Is intentionally specified backoffLimit as 0?

File: docs/upgrade.md (hunk 8 above)
Comment (shubham14bajpai) [reply docs/upgrade.md:10]: No it supports 1.10.0 as well

File: examples/migrate/cstor-volume-migration.yaml (hunk 9 above)
Comment (shubham14bajpai) [reply examples/migrate/cstor-volume-migration.yaml:36]: No specific reason just to get all the errors from different job pods

File: docs/migration.md (hunk 7 above)
Comment (shubham14bajpai) [reply docs/migration.md]: Done

**Comment by mittachaitu** [comment]
Can we add below note for rancher based cluster:

Note: If the Kubernetes cluster is on rancher and iscsi is running inside the kubelet container then it is mandatory to install iscsi service on the nodes and add extra binds to the kubelet container as mentioned [here](https://docs.openebs.io/docs/next/prerequisites.html#rancher).

**Comment by shubham14bajpai** [comment]
> Can we add below note for rancher based cluster:
> 
> Note: If the Kubernetes cluster is on rancher and iscsi is running inside the kubelet container then it is mandatory to install iscsi service on the nodes and add extra binds to the kubelet container as mentioned [here](https://docs.openebs.io/docs/next/prerequisites.html#rancher).

Added

**Pull request merged**